### DISCUSSION TITLE
VS Code appends .js to auto imports

### DISCUSSION THREAD

**Discussion by thefrana** [epicweb-dev/epic-stack#773] [Q&A]
Hi,
I was wondering if this behavior is normal. Since Typescript v5.2 (definitely on current v5.4) VS Code automatically add `.js` extension to all auto imports. I cannot change this through any VS Code or Typescript config. I would expect imports with no extension or with `.ts` extension. The main problem with `.js` extension is that it can sometimes break the build which is definitely not ideal.

Do you have any idea what can be causing this or how can I resolve it?

**Comment by kentcdodds** [comment]
I've noticed this happening recently as well. It's annoying though it doesn't actually negatively impact the functionality.

I would love a fix for this. I like keeping the extensions because I think it mimics native ESM better (which requires file extensions when referencing a file directly), but I could be convinced to just not bother I think.

I welcome anyone to dive in and see what can be done about this. I would definitely prefer to have the actual file extension be the one used in the code if possible.

**Reply by ceopaludetto** [reply]
Maybe you're using `"module": "preserve"` in `tsconfig.json`, I've noticed that if you use `"moduleResolution": "bundler"` and `"moduleDetection": "force"` you will get extensionless files, but if you add `"allowImportingTsExtensions": true` you will get `.tsx?` extensions as well.

So the recipe is:

`.js` extensions:
```json
{
  "compilerOptions": {
    "module": "preserve"
  }
}
```

extensionless:
```json
{
  "compilerOptions": {
    "moduleResolution": "bundler",
    "moduleDetection": "force"
  }
}
```

`.ts` and `.tsx` extensions:
```json
{
  "compilerOptions": {
    "moduleResolution": "bundler",
    "moduleDetection": "force",
    "allowImportingTsExtensions": true
  }
}
```

**Reply by ceopaludetto** [reply]
This also works on different editors (tested on zed 0.139.3), so maybe is more related to the TypeScript and the LSP itself instead of vscode configurations

**Reply by kentcdodds** [reply]
Sadly this isn't working for me (even after an editor restart)

https://github.com/epicweb-dev/epic-stack/assets/1500684/92c5611a-ebf2-4de3-96df-9925d5f1c1c2

**Reply by ceopaludetto** [reply]
Try this vscode configurations as well:

```jsonc
{
  // Javascript
  "javascript.preferences.importModuleSpecifier": "shortest",
  "javascript.preferences.importModuleSpecifierEnding": "minimal",

  // Typescript
  "typescript.preferences.importModuleSpecifier": "shortest",
  "typescript.preferences.importModuleSpecifierEnding": "minimal"
}
```

The behavior here is really inconsistent, if I type the import manually i get `.ts` and `.tsx` extensions, but if I try to auto import I get extensionless


https://github.com/epicweb-dev/epic-stack/assets/30508218/7e4f8ba5-b3c1-4069-97d2-410aab120bb2

**Reply by ceopaludetto** [reply]
Ok, using the following configuration combined with `moduleResolution`, `moduleDetection` and `allowImportingTsExtensions`:

```jsonc
  // Javascript
  "javascript.preferences.importModuleSpecifier": "shortest",
  "javascript.preferences.importModuleSpecifierEnding": "js",

  // Typescript
  "typescript.preferences.importModuleSpecifier": "shortest",
  "typescript.preferences.importModuleSpecifierEnding": "js",
```

I get `.ts` and `.tsx` extensions from both writing and auto importing

https://github.com/epicweb-dev/epic-stack/assets/30508218/c5093012-22b4-444a-870b-792e0c99cd44

**Reply by kentcdodds** [reply]
Thanks for working on this! I think this is working for relative paths, but for anything coming from the `compilerOptions.paths` config I'm not getting the behavior I want.

https://github.com/epicweb-dev/epic-stack/assets/1500684/522ad3aa-ca2f-44fe-a1a5-114d839a13e2

**Reply by ceopaludetto** [reply]
Here it works even for `compilerOptions.paths`, but I don't know why. I will try to figure out tonight

**Reply by kentcdodds** [reply]
Maybe there's something else about [my config](https://github.com/kentcdodds/ama/issues/406#issuecomment-391764106) or the Epic Stack that's causing the issue. Thanks for spending time working on this!

**Reply by thefrana** [reply]
I also tried to dig into this problem and I have not found a complete solution. There was some discussion in VS Code repo about adding `"typescript.preferences.importModuleSpecifierEnding": "ts"` option.

**Reply by kentcdodds** [reply]
I wonder if the problem has something to do with the `imports` config in `package.json` 🤔

**Reply by Klukies** [reply]
I'm pretty sure that this is in fact related to the `imports` config in `package.json` like Kent mentioned. I've been trying to get imports working without paths in tsconfig as this feature is now natively supported in TypeScript since [5.4.0](https://devblogs.microsoft.com/typescript/announcing-typescript-5-4/#auto-import-support-for-subpath-imports). However, having this enabled will always add a `.js` extensions regardless of the config I apply.

I personally don't agree with having file extensions, due to the fact that I think barrel files are still usefull in some situations and importing from an `index.(js|ts)` file feels weird. The only way to get this behaviour to work is to (sadly) remove the `imports` config in `package.json`

**Reply by kentcdodds** [reply]
That sounds like a bug in TypeScript then. Anyone want to open an issue (or find one to up vote)?

**Reply by PabloCorso** [reply]
I'm trying the following but not sure if the `imports` is actually needed or if this could be a proper workaround:

1. Remove `imports` from `package.json`.
2. Add `resolve.alias` in `vite.config.ts`:

```js
import path from 'node:path';

export default defineConfig({
  resolve: {
    alias: {
      '#app': path.resolve(__dirname, 'app'),
      '#tests': path.resolve(__dirname, 'tests'),
      '@': path.resolve(__dirname, 'app/components/ui/icons'),
    },
  },
};
```

3. Import suggests aliases with proper extensions with this workspace settings at `.vscode/settings.json`:

```json
{
  "typescript.preferences.importModuleSpecifier": "non-relative",
  "typescript.preferences.importModuleSpecifierEnding": "js"
}
```

**Comment by theskillwithin** [comment]
For me I have a similar strange behavior.  all imports work as expected in Remix.  

however when I run Playwright, it wants all js imports in my app thats shared with it, to all the sudden to have a .js extension

```
Error: Cannot find module '/node_modules/lodash/groupBy' imported from /app/utils/question.ts
Did you mean to import "lodash/groupBy.js"?
````

**Reply by jacobparis** [reply]
if tsconfig#moduleResolution is set to ESNext it will ask for extensions, if it's set to Bundler then it's ok without them

**Comment by thefrana** [comment]
VS Code's import resolution behavior seems inconsistent. When resolving relative path imports, it omits the .js extension. However, for imports using aliases (e.g., #app), it includes the .js extension. I've tried adjusting the "typescript.preferences.importModuleSpecifierEnding" setting, but none of the options achieve consistent behavior across both import types. I am not sure if this behavior was caused by some update or just did not notice before.

**Reply by Noriller** [reply]
I think it might be related to this:

https://www.totaltypescript.com/relative-import-paths-need-explicit-file-extensions-in-ecmascript-imports

**Comment by shawtung** [comment]
May be due to `allowImportingTsExtensions` not set true in tsconfig.

**Reply by thefrana** [reply]
Turns out the problem is most likely in the combination of VS Code version, its settings, TypeScript version and tsconfig. We eventually got some working combination and we pray it will stay that way.
